common/prove/witness_helper.go

```diff
@@ -597,3 +605,158 @@ func (w *WitnessHelper) constructSimpleWitnessInfo(oTx *tx.Tx) (
 	}
 	return accountKeys, accountWitnessInfo, nftWitnessInfo, nil
 }
+
+func (w *WitnessHelper) ConstructGasWitness(block *block.Block) (cryptoGas *GasWitness, stateRoot []byte, err error) {
+	var gas *circuit.Gas
+
+	needGas := false
+	gasBefore := make(map[int64]*big.Int)
+	gasChanges := make(map[int64]*big.Int)
+	for _, assetId := range types.GasAssets {
+		gasChanges[assetId] = types.ZeroBigInt
+	}
+	for _, tx := range block.Txs {
+		if types.IsL2Tx(tx.TxType) {
+			needGas = true
+			for _, txDetail := range tx.TxDetails {
+				if txDetail.IsGas {
+					assetBefore, err := types.ParseAccountAsset(txDetail.Balance)
+					if err != nil {
+						return nil, nil, err
+					}
+					if _, ok := gasBefore[assetBefore.AssetId]; !ok {
+						gasBefore[assetBefore.AssetId] = assetBefore.Balance
+					}
+					assetDelta, err := types.ParseAccountAsset(txDetail.BalanceDelta)
+					if err != nil {
+						return nil, nil, err
+					}
+					gasChanges[assetDelta.AssetId] = ffmath.Add(gasChanges[assetDelta.AssetId], assetDelta.Balance)
+				}
+			}
+		}
+	}
+
+	emptyAssetTree, err := tree.NewMemAccountAssetTree()
+	if err != nil {
+		return nil, nil, err
+	}
+	if !needGas { // no need of gas for this block
+		accountInfoBefore := cryptoTypes.EmptyGasAccount(LastAccountIndex, tree.NilAccountAssetRoot)
+		accountMerkleProofs, err := w.accountTree.GetProof(LastAccountIndex)
+		if err != nil {
+			return nil, nil, err
+		}
+		merkleProofsAccountBefore, err := SetFixedAccountArray(accountMerkleProofs)
+		if err != nil {
+			return nil, nil, err
+		}
+		merkleProofsAccountAssetsBefore := make([][AssetMerkleLevels][]byte, 0)
+		for i := 0; i < len(types.GasAssets); i++ {
+			accountInfoBefore.AssetsInfo = append(accountInfoBefore.AssetsInfo, cryptoTypes.EmptyAccountAsset(LastAccountAssetId))
+			assetMerkleProof, err := emptyAssetTree.GetProof(LastAccountAssetId)
+			if err != nil {
+				return nil, nil, err
+			}
+			merkleProofsAccountAssetBefore, err := SetFixedAccountAssetArray(assetMerkleProof)
+			if err != nil {
+				return nil, nil, err
+			}
+			merkleProofsAccountAssetsBefore = append(merkleProofsAccountAssetsBefore, merkleProofsAccountAssetBefore)
+		}
+		gas = &circuit.Gas{
+			GasAssetCount:                   len(types.GasAssets),
+			AccountInfoBefore:               accountInfoBefore,
+			MerkleProofsAccountBefore:       merkleProofsAccountBefore,
+			MerkleProofsAccountAssetsBefore: merkleProofsAccountAssetsBefore,
+		}
+	} else {
+		// gas account
+		gasAccountInfo, err := w.accountModel.GetConfirmedAccountByIndex(types.GasAccount)
+		if err != nil {
+
+			return nil, nil, err
+		}
+
+		pk, err := common2.ParsePubKey(gasAccountInfo.PublicKey)
+		if err != nil {
+			return nil, nil, err
+		}
+		accountInfoBefore := &cryptoTypes.GasAccount{
+			AccountIndex:    gasAccountInfo.AccountIndex,
+			AccountNameHash: common.FromHex(gasAccountInfo.AccountNameHash),
+			AccountPk:       pk,
+			Nonce:           gasAccountInfo.Nonce,
```

Comment:
The gasAccountInfo is got from the database, if the gas account itself has sent some txs in this block, this nonce maybe not correct here?

Reply:
yes, i will see how to fix this.

Reply:
I think the account before of gas account, should be profiled after the last transaction been executed.

Reply:
fixed, tracking gas account changes when re-run in witness